io-engine/src/bdev/ftl.rs

```diff
@@ -0,0 +1,310 @@
+                uri: uri.to_string(),
+                message: format!(
+                    "Could not percent decode cbdev_uri sub-uri - {}",
+                    e
+                ),
+            })?
+            .to_string();
+
+        // Device parameter checking for bbdev and cbdev are done in
+        // bdev_ftl_create_bdev
+        reject_unknown_parameters(uri, parameters)?;
+
+        Ok(Self {
+            name: uri.path()[1 ..].into(),
+            alias: uri.to_string(),
+            uuid,
+            bbdev_uri,
+            cbdev_uri,
+        })
+    }
+}
+
+impl GetName for Ftl {
+    fn get_name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+pub extern "C" fn ftl_bdev_init_fn_cb(
+    _ptr: *const ftl_bdev_info,
+    sender_ptr: *mut c_void,
+    errno: i32,
+) {
+    info!("{:?}: ftl_bdev_init_fn_cb", errno);
+    let sender = unsafe {
+        Box::from_raw(sender_ptr as *mut oneshot::Sender<ErrnoResult<()>>)
+    };
+    sender
+        .send(errno_result_from_i32((), errno))
+        .expect("done callback receiver side disappeared");
+}
+
+#[async_trait(?Send)]
+impl CreateDestroy for Ftl {
+    type Error = BdevError;
+
+    async fn create(&self) -> Result<String, Self::Error> {
+        if UntypedBdev::lookup_by_name(&self.name).is_some() {
+            return Err(BdevError::BdevExists {
+                name: self.name.clone(),
+            });
+        }
+
+        let ftl_dev_name = CString::new(self.name.clone()).unwrap();
+
+        debug!(
+            "{:?}: Creating ftl-bdev '{:?}' with bbdev {:?} and cbdev {:?}.",
+            self, ftl_dev_name, self.bbdev_uri, self.cbdev_uri
+        );
+
+        use std::ffi::CString;
+
+        let base_dev_name = bdev_create(&self.bbdev_uri).await?;
+        let cache_dev_name = bdev_create(&self.cbdev_uri).await?;
+
+        let (s, r) = oneshot::channel::<ErrnoResult<()>>();
+
+        let spdk_ftl_conf_size = mem::size_of::<spdk_ftl_conf>() as u64;
+        let mut ftl_conf = spdk_ftl_conf {
+            ..unsafe { mem::zeroed() }
+        };
+        unsafe { spdk_ftl_get_default_conf(&mut ftl_conf, spdk_ftl_conf_size) };
+        ftl_conf.name = ftl_dev_name.as_ptr() as *mut i8;
+        ftl_conf.base_bdev = base_dev_name.as_ptr() as *mut i8;
+        ftl_conf.cache_bdev = cache_dev_name.as_ptr() as *mut i8;
+        ftl_conf.fast_shutdown = true;
+        ftl_conf.verbose_mode = true;
+        ftl_conf.mode = spdk_ftl_mode::SPDK_FTL_MODE_CREATE as u32;
+
+        let errno = unsafe {
+            bdev_ftl_create_bdev(
+                &ftl_conf,
+                Some(ftl_bdev_init_fn_cb),
+                cb_arg(s),
+            )
+        };
+
+        if errno != 0 {
+            let err = BdevError::CreateBdevFailed {
+                source: Errno::from_raw(errno.abs()),
+                name: self.name.clone(),
+            };
+
+            error!("{:?} error: {}", self, err.verbose());
+
+            return Err(err);
```

Comment:
Shouldn't we destroy the bbdev and cbdev created earlier here, in case of ftl bdev creation failed?

Reply:
I agree :) Will address this tomorrow.

Reply:
Possibly though may be difficult as `Drop` is blocking. We could potentially just schedule the deletion to happen asynchronously, though this may become an awkward api.
Another question here is, what if the b and c devices are exist?  Presumably this fails on `bdev_create`?

Reply:
> Possibly though may be difficult as Drop is blocking. We could potentially just schedule the deletion to happen asynchronously, though this may become an awkward api.

Okay, then I will handle those cases individually :)

> Another question here is, what if the b and c devices are exist? Presumably this fails on bdev_create?

Yes it is going to fail, and I would propose to keep it like this because we don't want to accidentally reformat a drive that is already in use :)